embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c

```diff
@@ -125,99 +132,182 @@ void playMorsePhrase (const char* morsePhraseToPlay) {
 		return;
 }
 
+enum ButtonState { PRESSED, RELEASED };
 
-unsigned char playFlag = 0;
-unsigned char button = 0;
+typedef struct {
+	unsigned char pin : 3;
+	char keyChar;
+} Button;
 
-unsigned char button1 = ~(1 << PB6);
-unsigned char button2 = ~(1 << PB1);
-unsigned char button3 = ~(1 << PB4);
+Button buttons[] = {
+	{ PB0, '1'},
+	{ PB1, '2'},
+	{ PB2, '3'},
+	{ PB3, '4'},
+	{ PB4, '5'},
+	{ PB5, '6'},
+	{ PB6, '7'},
+	{ PB7, '8'}
+};
 
-ISR (PCINT_vect) {
-	static unsigned char i = 0;
-	unsigned char data = PINB;
-	
-	// delay while noise finishes and clear interrupt flag
-	// to do not cause extra interrupt
-	_delay_ms(500);
-	EIFR |= 1 << PCIF;
+#define BTN_1_REL (RELEASED << PB0)
+#define BTN_2_REL (RELEASED << PB1)
+#define BTN_3_REL (RELEASED << PB2)
+#define BTN_4_REL (RELEASED << PB3)
+#define BTN_5_REL (RELEASED << PB4)
+#define BTN_6_REL (RELEASED << PB5)
+#define BTN_7_REL (RELEASED << PB6)
+#define BTN_8_REL (RELEASED << PB7)
+
+#define ALL_RELEASED	BTN_1_REL | BTN_2_REL | BTN_3_REL | BTN_4_REL | \
+						BTN_5_REL | BTN_6_REL | BTN_7_REL | BTN_8_REL
+
+uint32_t debounceTime = 100;
+
+// cyclic buffer implementation
+typedef struct {
+	unsigned char pos;
+	unsigned char size;
+	char buff[MAX_PASS_LEN];
+}CyclicBuffer;
+
+// functions to handle cyclic buffer
+void initCyclicBuffer(CyclicBuffer *buff);
+void addCharToCyclicBuffer(CyclicBuffer *buff, const char ch);
+bool findSequenceInCyclicBuffer(CyclicBuffer *buff, const char *seq);
+
+CyclicBuffer cyclicBuffer;
+
+void initCyclicBuffer(CyclicBuffer *buff) {
+	buff->pos = 0;
+	buff->size = MAX_PASS_LEN;
```

Comment:
Agree, about resource waste. At that time i thought that it will make code more readable(at least for me), as buffer size is inside buffer struct, and not defined anywhere else.
So, not to waste resources it is possible to determine size from `MAX_PASS_LEN` macro and so `CyclicBuffer` struct may become:
```c
typedef struct {
	unsigned char pos;
	char buff[MAX_PASS_LEN];
}CyclicBuffer;
```
And everywhere in code use MAX_PASS_LEN as common buffer size.

What are the benefits of the code you introduced? 
Does it makes `CyclicBuffer` init routine more universal?

Reply:
I'm so glad to see counter-questions from you!
TL;DR - I was wrong. Long story below:

My intended reason behind that code: de-couple buffer size from structure definition so that I could write unit-tests for different buffer sizes. And to actually do that I must omit buffer size in structure definition:
```C
typedef struct {
	unsigned char size;
	unsigned char pos;
	char buff[];
} CyclicBuffer;
```
This works nice for run-time allocation:
```C
void doSomeStuff(unsigned char buffSize) {
     CyclicBuffer *cb = initCyclicBuffer(malloc(sizeof(CyclicBuffer) + buffSize), buffSize);
}
```
**BUT** to allocate buffer **with specific size** at **compile-time** becomes a pain:
```C
char cb2_data[MAX_PASS_LEN + sizeof(CyclicBuffer)];
CyclicBuffer * cb2 = (CyclicBuffer *)cb2_data;
void main() {
    initCyclicBuffer(cb2, sizeof(cb2_data));
```
And now it became really inefficient - I'm forced to use pointer.
How about using union to pack CyclicBuffer and data buffer in one space?
```C
union {
	CyclicBuffer h;
	char data[MAX_PASS_LEN + sizeof(CyclicBuffer)];
} cb2;
void main() {
    initCyclicBuffer(&cb2.h, sizeof(cb2));
```
This is kind of better, but still - I force myself to use magic "h" (stand for "header") union member. And one line definiton is now bloated to to 4 lines 
Some other reasons I was going to use to justify my megacode code:
- allows many CyclicBuffer-s with different size each.
- create buffers at run-time.
- re-use any chunk of memory as CyclicBuffer.

Is this needed for this particular project - no. 
Maybe my code could be justified in larger project, where many buffers with different sizes needs to be allocated dynamically, but when one global buffer is needed - it is just a bloat.

p.s. In programming such buffers often are called "Circular" or "Round" buffers. Write position is called "head" and read position is called "tail" (worm/snake analogy).